File: tutorials/phase-2-backend/04-r-integration/src/r-scripts/mutation_analysis.R
# Mutation Analysis Script
# Statistical analysis of mutation patterns

# Load required libraries
if (!require("stats")) install.packages("stats", repos="http://cran.r-project.org")

# Function: Chi-square test for mutation enrichment
mutation_enrichment <- function(input_file, output_file = NULL) {
  data <- read.csv(input_file)
  
  # Get cancer type columns
  cancer_types <- c("breast", "lung", "colon", "ovarian", "pancreatic", "melanoma")
  
  results <- list()
  
  for (i in 1:nrow(data)) {
    gene <- data$gene[i]
    observed <- as.numeric(data[i, cancer_types])
    total <- data$total_samples[i]
    
    # Calculate expected frequencies (assuming uniform distribution)
    expected <- rep(total / length(cancer_types), length(cancer_types))
    
    # Chi-square test
    chi_result <- chisq.test(observed, p = rep(1/length(cancer_types), length(cancer_types)))
    
    # Find enriched cancer types
    enrichment_ratio <- observed / expected
    enriched <- cancer_types[enrichment_ratio > 1.5]
    depleted <- cancer_types[enrichment_ratio < 0.5]
    
    results[[gene]] <- list(
      gene = gene,
      total_mutations = sum(observed),
      by_cancer_type = setNames(as.list(observed), cancer_types),
      enrichment_ratio = setNames(as.list(round(enrichment_ratio, 2)), cancer_types),
      chi_square = chi_result$statistic,
      pvalue = chi_result$p.value,
      significant = chi_result$p.value < 0.05,
      enriched_in = enriched,
      depleted_in = depleted
    )
  }
  
  result <- list(
    analysis_type = "mutation_enrichment",
    n_genes = nrow(data),
    cancer_types = cancer_types,
    results = results
  )
  
  if (!is.null(output_file)) {
    json_output <- jsonlite::toJSON(result, auto_unbox = TRUE, pretty = TRUE)
    writeLines(json_output, output_file)
  }
  
  return(result)
}

# Function: Mutual exclusivity analysis
mutual_exclusivity <- function(input_file, output_file = NULL) {
  data <- read.csv(input_file)
  
  # Get cancer type columns
  cancer_types <- c("breast", "lung", "colon", "ovarian", "pancreatic", "melanoma")
  
  # Calculate correlation between mutation patterns
  mutation_matrix <- as.matrix(data[, cancer_types])
  rownames(mutation_matrix) <- data$gene
  
  # Normalize by total samples
  mutation_freq <- mutation_matrix / data$total_samples
  
  # Calculate correlation
  cor_matrix <- cor(t(mutation_freq))
  
  # Find mutually exclusive pairs (negative correlation)
  pairs <- list()
  genes <- data$gene
  
  for (i in 1:(length(genes) - 1)) {
    for (j in (i + 1):length(genes)) {
      correlation <- cor_matrix[i, j]
      pairs[[length(pairs) + 1]] <- list(
        gene1 = genes[i],
        gene2 = genes[j],
        correlation = round(correlation, 3),
        relationship = ifelse(correlation < -0.3, "mutually_exclusive",
                             ifelse(correlation > 0.3, "co_occurring", "independent"))
      )
    }
  }
  
  # Sort by absolute correlation
  pairs <- pairs[order(-sapply(pairs, function(x) abs(x$correlation)))]
  
  result <- list(
    analysis_type = "mutual_exclusivity",
    n_genes = nrow(data),
    pairs = pairs[1:20]  # Top 20 pairs
  )
  
  if (!is.null(output_file)) {
    json_output <- jsonlite::toJSON(result, auto_unbox = TRUE, pretty = TRUE)
    writeLines(json_output, output_file)
  }
  
  return(result)
}

# Function: Cancer type clustering by mutation profile
cancer_clustering <- function(input_file, output_file = NULL) {
  data <- read.csv(input_file)
  
  # Get cancer type columns
  cancer_types <- c("breast", "lung", "colon", "ovarian", "pancreatic", "melanoma")
  
  # Create matrix with cancer types as rows
  mutation_matrix <- t(as.matrix(data[, cancer_types]))
  colnames(mutation_matrix) <- data$gene
  
  # Normalize by row (cancer type)
  row_sums <- rowSums(mutation_matrix)
  mutation_freq <- mutation_matrix / row_sums
  
  # Hierarchical clustering
  dist_matrix <- dist(mutation_freq, method = "euclidean")
  hclust_result <- hclust(dist_matrix, method = "ward.D2")
  
  # Get dendrogram order
  order <- hclust_result$order
  
  result <- list(
    analysis_type = "cancer_clustering",
    cancer_types = cancer_types,
    dendrogram_order = cancer_types[order],
    height = hclust_result$height,
    merge = hclust_result$merge
  )
  
  if (!is.null(output_file)) {
    json_output <- jsonlite::toJSON(result, auto_unbox = TRUE, pretty = TRUE)
    writeLines(json_output, output_file)
  }
  
  return(result)
}

# Command line interface
args <- commandArgs(trailingOnly = TRUE)
if (length(args) >= 2) {
  analysis_type <- args[1]
  input_file <- args[2]
  output_file <- if (length(args) >= 3) args[3] else NULL
  
  if (analysis_type == "enrichment") {
    result <- mutation_enrichment(input_file, output_file)
  } else if (analysis_type == "exclusivity") {
    result <- mutual_exclusivity(input_file, output_file)
  } else if (analysis_type == "clustering") {
    result <- cancer_clustering(input_file, output_file)
  }
  
  if (is.null(output_file)) {
    cat(jsonlite::toJSON(result, auto_unbox = TRUE, pretty = TRUE))
  }
}

File: tutorials/phase-2-backend/04-r-integration/src/examples/mutationAnalysis.js
/**
 * Mutation Analysis Example
 * Demonstrates mutation enrichment and pattern analysis
 */

import { fileURLToPath } from 'url';
import { dirname, join } from 'path';
import { promises as fs } from 'fs';
import jsStats from '../services/jsStats.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);
const DATA_DIR = join(__dirname, '..', '..', 'data');

async function mutationAnalysisExample() {
  console.log(`
═══════════════════════════════════════════════════════════
  Mutation Analysis Example
═══════════════════════════════════════════════════════════
`);

  // Load mutation frequency data (skip comment lines)
  const csvContent = await fs.readFile(join(DATA_DIR, 'mutation_frequency.csv'), 'utf-8');
  const lines = csvContent.trim().split('\n').filter(line => line.trim() && !line.startsWith('#'));
  const headers = lines[0].split(',').map(h => h.trim());
  
  const data = lines.slice(1).map(line => {
    const values = line.split(',');
    const obj = {};
    headers.forEach((h, i) => {
      obj[h] = h === 'gene' ? values[i].trim() : parseInt(values[i]);
    });
    return obj;
  });

  const cancerTypes = ['breast', 'lung', 'colon', 'ovarian', 'pancreatic', 'melanoma'];

  console.log('📊 Data Summary:');
  console.log('─'.repeat(60));
  console.log(`  Genes analyzed: ${data.length}`);
  console.log(`  Cancer types: ${cancerTypes.join(', ')}`);

  // Mutation Enrichment Analysis
  console.log('\n📍 Mutation Enrichment by Cancer Type:');
  console.log('─'.repeat(60));
  
  for (const row of data.slice(0, 5)) {
    const observed = cancerTypes.map(ct => row[ct]);
    const total = row.total_samples;
    const expected = cancerTypes.map(() => total / cancerTypes.length);
    
    const chiResult = jsStats.chiSquareTest(observed, expected);
    
    // Find enriched types
    const enriched = cancerTypes.filter((ct, i) => observed[i] / expected[i] > 1.5);
    
    console.log(`\n  ${row.gene}:`);
    console.log(`    Total mutations: ${observed.reduce((a, b) => a + b, 0)}`);
    console.log(`    Chi-square: ${chiResult.statistic.toFixed(2)}, p-value: ${chiResult.pvalue.toExponential(2)}`);
    console.log(`    Enriched in: ${enriched.length > 0 ? enriched.join(', ') : 'none'}`);
    
    // Show breakdown
    console.log('    Breakdown:');
    cancerTypes.forEach((ct, i) => {
      const ratio = (observed[i] / expected[i]).toFixed(2);
      const bar = '█'.repeat(Math.min(Math.round(observed[i] / 20), 20));
      const marker = ratio > 1.5 ? '↑' : ratio < 0.5 ? '↓' : ' ';
      console.log(`      ${ct.padEnd(11)}: ${observed[i].toString().padStart(3)} ${bar} ${marker}`);
    });
  }

  // Mutation Pattern Heatmap Data
  console.log('\n📍 Mutation Frequency Heatmap:');
  console.log('─'.repeat(60));
  console.log('\n  Mutation frequency (per 1000 samples):');
  console.log('  ' + ''.padStart(10) + cancerTypes.map(ct => ct.slice(0, 6).padStart(8)).join(''));
  console.log('  ' + '─'.repeat(10 + cancerTypes.length * 8));
  
  for (const row of data) {
    const values = cancerTypes.map(ct => {
      const freq = (row[ct] / row.total_samples * 1000).toFixed(0);
      return freq.padStart(8);
    });
    console.log(`  ${row.gene.padEnd(10)}${values.join('')}`);
  }

  // Mutual Exclusivity Analysis
  console.log('\n📍 Mutual Exclusivity / Co-occurrence Analysis:');
  console.log('─'.repeat(60));
  
  const genes = data.map(d => d.gene);
  const pairs = [];
  
  for (let i = 0; i < genes.length - 1; i++) {
    for (let j = i + 1; j < genes.length; j++) {
      const x = cancerTypes.map(ct => data[i][ct]);
      const y = cancerTypes.map(ct => data[j][ct]);
      const cor = jsStats.correlation(x, y);
      
      pairs.push({
        gene1: genes[i],
        gene2: genes[j],
        correlation: cor,
        relationship: cor < -0.3 ? 'Mutually exclusive' :
                     cor > 0.3 ? 'Co-occurring' : 'Independent'
      });
    }
  }
  
  // Sort by absolute correlation
  pairs.sort((a, b) => Math.abs(b.correlation) - Math.abs(a.correlation));
  
  console.log('\n  Top Gene Pairs by Correlation:');
  console.log('  ┌──────────┬──────────┬────────────┬───────────────────────┐');
  console.log('  │  Gene 1  │  Gene 2  │    Corr    │     Relationship      │');
  console.log('  ├──────────┼──────────┼────────────┼───────────────────────┤');
  
  for (const p of pairs.slice(0, 10)) {
    const marker = p.correlation > 0 ? '🟢' : '🔴';
    console.log(`  │ ${p.gene1.padEnd(8)} │ ${p.gene2.padEnd(8)} │ ${p.correlation.toFixed(3).padStart(10)} │ ${marker} ${p.relationship.padEnd(19)} │`);
  }
  console.log('  └──────────┴──────────┴────────────┴───────────────────────┘');

  // Cancer Type Clustering
  console.log('\n📍 Cancer Type Similarity (by mutation profile):');
  console.log('─'.repeat(60));
  
  // Calculate distance between cancer types
  const cancerPairs = [];
  for (let i = 0; i < cancerTypes.length - 1; i++) {
    for (let j = i + 1; j < cancerTypes.length; j++) {
      const x = data.map(d => d[cancerTypes[i]] / d.total_samples);
      const y = data.map(d => d[cancerTypes[j]] / d.total_samples);
      const cor = jsStats.correlation(x, y);
      
      cancerPairs.push({
        cancer1: cancerTypes[i],
        cancer2: cancerTypes[j],
        similarity: cor
      });
    }
  }
  
  cancerPairs.sort((a, b) => b.similarity - a.similarity);
  
  console.log('\n  Most Similar Cancer Types:');
  for (const p of cancerPairs.slice(0, 5)) {
    const bar = '█'.repeat(Math.round((p.similarity + 1) * 10));
    console.log(`    ${p.cancer1} ↔ ${p.cancer2}: ${p.similarity.toFixed(3)} ${bar}`);
  }

  console.log(`
═══════════════════════════════════════════════════════════
  Analysis Complete
═══════════════════════════════════════════════════════════
`);
}

mutationAnalysisExample();
